### PR TITLE
Provisioning: datasources auto deletion 

### LINKED ISSUE #12875: Provisioning datasources should not require manual handling of deletions
Currently, you have to manually add a section to delete data sources, this should all happen automatically.

### PR DESCRIPTION
**Pull request by mikelv92**
**What is this feature?**

This PR makes it possible to delete a provisioned datasource by simply deleting it from the provisioning yaml file, or by deleting the yaml file.

**Why do we need this feature?**

Currently, you have to manually add a section to delete data sources, this should all happen automatically. 

**Who is this feature for?**

This feature is for users that provision datasources in yaml files and want to delete them automatically.

**Which issue(s) does this PR fix?**:

Fixes #12875

**Special notes for your reviewer:**

Steps to test it:
- Create a data source provisioning file using api version 1 and add `prune: true` in the file
- Create a provisioned datasource
- Start grafana: The datasource will show up in the list of datasources
- Delete the datasource from the yaml file or delete the provisioning yaml file
- Restart grafana: The datasource should also be deleted

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by mikelv92** [comment]
@wbrowne @andresmgot @oshirohugo @papagian @diegommm @yangkb09 Could you please review this PR? :) Or maybe it is no longer needed?

**Comment by andresmgot** [comment]
@grafana/backend-platform can correct me if I am wrong but I don't think we want to do this anymore. This would unexpectedly start deleting datasources from many instances with no path to recover them. My +1 would be to keep the current behavior and _only_ delete datasources that are explicitly marked for deletion.

**Comment by bjornfor** [comment]
> This would unexpectedly start deleting datasources from many instances with no path to recover them. 

Who provisions from a _temporary_ file? In my setups, the provisioning file is the single source of truth, and it's in git.

And I've been waiting for this fix for years.

**Comment by andresmgot** [comment]
> Who provisions from a _temporary_ file? In my setups, the provisioning file is the single source of truth, and it's in git.

I understand your use case but there is also the use case of people creating datasources through provisioning files that don't  expect them to be deleted when provisioning other datasources from a file that happens to be in the same path.

If we want to move forward with this, the default behavior should remain as it is, to avoid breaking changes. My suggestion is to include a new parameter (e.g. `prune`), that if enabled, will delete the datasources if they no longer appear:

```yml
apiVersion: 1

prune: true
datasources:
  - name: PostgreSQL
    type: postgres
    ...
```

**Comment by mikelv92** [comment]
@andresmgot Could you please review again if you have time? 🙂

File: pkg/services/provisioning/datasources/datasources.go [hunk 1]
```diff
@@ -153,6 +154,28 @@ func (dc *DatasourceProvisioner) applyChanges(ctx context.Context, configPath st
...
+	staleProvisionedDataSources := []*deleteDatasourceConfig{}
+
+	for _, prunableProvisionedDataSource := range prunableProvisionedDataSources {
+		key := DataSourceMapKey{
+			OrgId: prunableProvisionedDataSource.OrgID,
+			Name:  prunableProvisionedDataSource.Name,
+		}
+		if _, ok := willExistAfterProvisioning[key]; !ok {
+			staleProvisionedDataSources = append(staleProvisionedDataSources, &deleteDatasourceConfig{OrgID: prunableProvisionedDataSource.OrgID, Name: prunableProvisionedDataSource.Name})
+			willExistAfterProvisioning[key] = false
```
Comment (andresmgot) [inline pkg/services/provisioning/datasources/datasources.go:171]: You are not using the provision path in order to differ if a datasource should be deleted or not. I don't think it's actually needed. You can just use `is_prunable` and remove `provisionedFrom` from the model. A datasource marked to be pruned can be deleted whenever it does not appear anymore, regardless of the provision file path.

File: pkg/services/sqlstore/migrations/datasource_mig.go [hunk 2]
```diff
@@ -134,4 +134,12 @@ func addDataSourceMigration(mg *Migrator) {
 
 	mg.AddMigration("add unique index datasource_org_id_is_default", NewAddIndexMigration(tableV2, &Index{
 		Cols: []string{"org_id", "is_default"}}))
+
+	mg.AddMigration("Add provisioned_from column", NewAddColumnMigration(tableV2, &Column{
+		Name: "provisioned_from", Type: DB_NVarchar, Length: 40, Nullable: true,
+	}))
+
+	mg.AddMigration("Add is_prunable column", NewAddColumnMigration(tableV2, &Column{
+		Name: "is_prunable", Type: DB_Bool, Nullable: false, Default: "0",
```
Comment (andresmgot) [inline pkg/services/sqlstore/migrations/datasource_mig.go:143]: I believe this could be nullable.

**Review by andresmgot — COMMENTED** [review]
Thanks @mikelv92. I have done a first round of review but we would need to get more CODEOWNERs eyes here.

**Comment by mikelv92** [comment]
Thanks for the review @andresmgot . I made the changes according to your comments :)

**Review by andresmgot — COMMENTED** [review]
Thanks, one comment here. Also, you have some[ lint issues](https://drone.grafana.net/grafana/grafana/170083/2/6) (you can run the check with `make lint-go`).

File: pkg/services/datasources/service/store.go [hunk 3]
```diff
@@ -343,6 +355,7 @@ func (ss *SqlStore) UpdateDataSource(ctx context.Context, cmd *datasources.Updat
 			ReadOnly:        cmd.ReadOnly,
 			Version:         cmd.Version + 1,
 			UID:             cmd.UID,
+			IsPrunable: 	 cmd.IsPrunable,
```
[the remark this answers is not in the document]
Comment (mikelv92) [reply pkg/services/datasources/service/store.go]: Thanks! I added it

File: pkg/services/datasources/models.go [hunk 4]
```diff
@@ -62,6 +62,7 @@ type DataSource struct {
 	SecureJsonData    map[string][]byte `json:"secureJsonData"`
 	ReadOnly          bool              `json:"readOnly"`
 	UID               string            `json:"uid" xorm:"uid"`
+	IsPrunable        bool              `xorm:"is_prunable"`
```
Comment (andresmgot) [inline pkg/services/datasources/models.go:65]: this needs to be ignored for swagger (as for example, `BasicAuthPassword`)

**Review by andresmgot — COMMENTED** [review]
LGTM with a comment, thanks!

**Review by marefr — CHANGES_REQUESTED** [review]
Thanks. Some minor things left, see comments. In addition, would be nice if you can update the [documentation](https://github.com/grafana/grafana/blob/main/docs/sources/administration/provisioning/index.md#data-sources) as well.

File: docs/sources/administration/provisioning/index.md [hunk 5]
```diff
@@ -81,6 +81,24 @@ If the data source already exists, Grafana reconfigures it to match the provisio
...
+datasources:
+  - name: PostgreSQL
+    type: postgres
+    ...
+```
+
+If the datasources are provisioned with this parameter, they can also be deleted by simply removing them from the YAML file or by deleting the YAML file.
+
+> [!NOTE]
+> This `prune` parameter is available only in apiVersion 1
```
Comment (andresmgot) [inline docs/sources/administration/provisioning/index.md:99]: It's more relevant to specify the Grafana version. Also the format is different:

```suggestion
{{% admonition type="note" %}}
The `prune` parameter is available in Grafana v11.1 and higher.
{{% /admonition %}}
```

File: pkg/services/datasources/models.go [hunk 6]
```diff
@@ -161,6 +163,7 @@ type AddDataSourceCommand struct {
 	JsonData        *simplejson.Json  `json:"jsonData"`
 	SecureJsonData  map[string]string `json:"secureJsonData"`
 	UID             string            `json:"uid"`
+	IsPrunable      bool
```
Comment (andresmgot) [inline pkg/services/datasources/models.go:166]: You also need to add the comment  `swagger:ignore` here and in `UpdateDataSourceCommand`

**Review by andresmgot — COMMENTED** [review]
Some documentation comments here, thanks!

File: docs/sources/administration/provisioning/index.md [hunk 7]
```diff
@@ -100,6 +107,10 @@ deleteDatasources:
   - name: Graphite
     orgId: 1
 
+# Mark provisioned datasources for deletion if they are no longer in a provisioning file.
+# It takes no effect if datasources are already listed in the deleteDatasources section.
+prune: true
```
Comment (jdbaldry) [inline docs/sources/administration/provisioning/index.md:112]:
```suggestion
# Mark provisioned data sources for deletion if they are no longer in a provisioning file.
# It takes no effect if data sources are already listed in the deleteDatasources section.
prune: true
```

File: pkg/services/datasources/datasources.go [hunk 8]
```diff
@@ -21,6 +21,9 @@ type DataSourceService interface {
 	// GetAllDataSources gets all datasources.
 	GetAllDataSources(ctx context.Context, query *GetAllDataSourcesQuery) (res []*DataSource, err error)
 
+	// GetPrunableProvisionedDataSources gets all provisioned datasources that can be pruned.
```
Comment (jdbaldry) [inline pkg/services/datasources/datasources.go:24]:
```suggestion
	// GetPrunableProvisionedDataSources gets all provisioned data sources that can be pruned.
```

File: pkg/services/datasources/fakes/fake_datasource_service.go [hunk 9]
```diff
@@ -45,6 +45,16 @@ func (s *FakeDataSourceService) GetAllDataSources(ctx context.Context, query *da
 	return s.DataSources, nil
 }
 
+func (s *FakeDataSourceService) GetPrunableProvisionedDataSources(ctx context.Context) (res []*datasources.DataSource, err error) {
+	var dataSources []*datasources.DataSource
+	for _, datasource := range s.DataSources {
+		if datasource.IsPrunable {
+			dataSources = append(dataSources, datasource)
+		}
+	}
```
Comment (jdbaldry) [inline pkg/services/datasources/fakes/fake_datasource_service.go:54]: nit: consistent with the two word data source in `s.DataSources`

```suggestion
	for _, dataSource := range s.DataSources {
		if dataSource.IsPrunable {
			dataSources = append(dataSources, dataSource)
		}
	}
```

File: pkg/services/datasources/service/store.go [hunk 10]
```diff
@@ -124,6 +125,16 @@ func (ss *SqlStore) GetDataSourcesByType(ctx context.Context, query *datasources
 	})
 }
 
+// GetPrunableProvisionedDataSources returns all datasources that can be pruned
```
Comment (jdbaldry) [inline pkg/services/datasources/service/store.go:128]:
```suggestion
// GetPrunableProvisionedDataSources returns all data sources that can be pruned
```

File: pkg/services/datasources/service/store_test.go [hunk 11]
```diff
@@ -489,5 +489,41 @@ func TestIntegrationDataAccess(t *testing.T) {
 			require.NoError(t, err)
 			require.Equal(t, 1, len(dataSources))
 		})
+
+		t.Run("Get prunable datasources", func(t *testing.T) {
```
Comment (jdbaldry) [inline pkg/services/datasources/service/store_test.go:493]:
```suggestion
		t.Run("Get prunable data sources", func(t *testing.T) {
```

File: pkg/services/provisioning/datasources/config_reader_test.go [hunk 12]
```diff
@@ -169,6 +171,39 @@ func TestDatasourceAsConfig(t *testing.T) {
 		require.Equal(t, len(store.updated), 1)
 	})
 
+	t.Run("Delete datasources when removing them from provision files", func(t *testing.T) {
```
Comment (jdbaldry) [inline pkg/services/provisioning/datasources/config_reader_test.go:174]:
```suggestion
	t.Run("Delete data sources when removing them from provision files", func(t *testing.T) {
```

**Review by jdbaldry — COMMENTED** [review]
I have limited suggestions for the Go code to comments and variable naming consistency rather and haven't actually reviewed the code.

**Review by andresmgot — APPROVED** [review]
LGTM! Thanks

**Review by jdbaldry — APPROVED** [review]
docs LGTM

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**